# ApexPlanet Internship – Task 1
## Sales Data Immersion & Wrangling

### Objective
To understand, assess, clean, transform, and prepare the sales dataset
for further analysis.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [12]:
df = pd.read_excel(r"C:\Users\HP\Downloads\ApexPlanet_Dataset.xlsx")

In [13]:
df.head()

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


In [26]:
df.shape

(1000, 12)

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       1000 non-null   object 
 1   Order_Date     1000 non-null   object 
 2   Customer_ID    1000 non-null   object 
 3   Customer_Name  1000 non-null   object 
 4   Age            980 non-null    float64
 5   Gender         1000 non-null   object 
 6   City           987 non-null    object 
 7   Product        1000 non-null   object 
 8   Category       1000 non-null   object 
 9   Quantity       1000 non-null   int64  
 10  Unit_Price     1000 non-null   float64
 11  Total_Sales    1000 non-null   float64
dtypes: float64(3), int64(1), object(8)
memory usage: 93.9+ KB


In [28]:
df.describe()

,Age,Quantity,Unit_Price,Total_Sales
count,980.000000,1000.000000,1000.000000,1000.000000
mean,41.360204,5.435000,25486.783410,139399.439650
std,13.822597,2.838632,14179.402361,114100.051546
min,18.000000,1.000000,145.780000,437.340000
25%,30.000000,3.000000,13895.722500,47066.632500
50%,41.000000,5.000000,25398.740000,108594.025000
75%,54.000000,8.000000,37512.382500,203722.882500
max,65.000000,10.000000,49997.530000,493677.500000


In [21]:
df.columns

Index(['Order_ID', 'Order_Date', 'Customer_ID', 'Customer_Name', 'Age',
       'Gender', 'City', 'Product', 'Category', 'Quantity', 'Unit_Price',
       'Total_Sales'],
      dtype='object')

## Checking Missing Values

In [23]:
df.isnull().sum()

Order_ID          0
Order_Date        0
Customer_ID       0
Customer_Name     0
Age              20
Gender            0
City             13
Product           0
Category          0
Quantity          0
Unit_Price        0
Total_Sales       0
dtype: int64

In [24]:
df.isnull().sum().sort_values(ascending= False)

Age              20
City             13
Order_ID          0
Order_Date        0
Customer_Name     0
Customer_ID       0
Gender            0
Product           0
Category          0
Quantity          0
Unit_Price        0
Total_Sales       0
dtype: int64

## Checking Duplicate Values

In [30]:
df.duplicated().sum()

np.int64(0)

In [31]:
df["Order_ID"].duplicated().sum()

np.int64(8)

In [32]:
duplicate_orders = df[df["Order_ID"].duplicated(keep=False)].sort_values("Order_ID")
duplicate_orders

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,NaN,Male,Pune,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,34.0,Male,Kolkata,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,23.0,Male,Bengaluru,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,24.0,Female,Pune,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,64.0,Male,Bengaluru,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,NaN,Male,NaN,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,31.0,Female,Mumbai,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,35.0,Male,Kolkata,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,62.0,Male,Bengaluru,Mobile,Electronics,1,34619.77,34619.77


In [33]:
duplicate_orders[["Order_ID", "Order_Date", "Customer_ID", "Customer_Name",
                  "Product", "Category", "Quantity", "Unit_Price", "Total_Sales"]]

,Order_ID,Order_Date,Customer_ID,Customer_Name,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,Mobile,Electronics,1,34619.77,34619.77


## change float to datetime (data types)

In [34]:
df["Order_Date"] = pd.to_datetime(df["Order_Date"])

### For changing data types create 
### "Calculated_Sales : quan x unitprice"
### "Sales_Match : tot(2)==cal_sales(2)"

In [46]:
df["Calculated_Sales"] = (df["Quantity"] * df["Unit_Price"])

In [47]:
df["Sales_Match"] = (df["Total_Sales"].round(2) == df["Calculated_Sales"].round(2))

In [48]:
df["Sales_Match"].value_counts()

Sales_Match
True    1000
Name: count, dtype: int64

### Now cleaned so remove that two Columns "Calculated_Sales" and "Sales_Match"

In [49]:
df.drop(columns=["Calculated_Sales", "Sales_Match"], inplace=True)

In [50]:
df.shape

(1000, 12)

### Check unique values in categorical columns

In [51]:
categorical_columns = ["Gender","City","Product","Category"]

for column in categorical_columns:
    print(f"\n--- {column} ---")
    print("Unique values:", df[column].nunique())
    print(df[column].unique())


--- Gender ---
Unique values: 2
['Female' 'Male']

--- City ---
Unique values: 8
['Bengaluru' 'Kolkata' 'Hyderabad' 'Patna' 'Mumbai' 'Delhi' 'Gaya' 'Pune'
 nan]

--- Product ---
Unique values: 6
['Rice' 'Book' 'Mobile' 'Laptop' 'Shoes' 'Chair']

--- Category ---
Unique values: 5
['Grocery' 'Education' 'Electronics' 'Fashion' 'Furniture']


### Categorical Data Quality Assessment

The categorical columns were checked for unique values and formatting consistency. Gender contains two consistent values (Male and Female), while Product and Category contain consistent values without apparent spelling or capitalization issues. City contains eight valid city values, along with 13 missing records that require treatment during the cleaning stage.

In [52]:
text_columns = ["Order_ID","Customer_ID","Customer_Name","Gender","City","Product","Category"]

for column in text_columns:
    spaces = df[column].dropna().astype(str).str.strip().ne(
        df[column].dropna().astype(str)
    ).sum()
    
    print(f"{column}: {spaces} values with extra spaces")

Order_ID: 0 values with extra spaces
Customer_ID: 0 values with extra spaces
Customer_Name: 0 values with extra spaces
Gender: 0 values with extra spaces
City: 0 values with extra spaces
Product: 0 values with extra spaces
Category: 0 values with extra spaces


### Text Formatting Validation

Leading and trailing spaces were checked across the identifier and categorical text columns. No extra spaces were detected in any of the columns, indicating consistent basic text formatting.

In [53]:
print("Data type:", df["Order_Date"].dtype)
print("\nSample dates:")
print(df["Order_Date"].head(10))

print("\nNumber of unique dates:", df["Order_Date"].nunique())

Data type: datetime64[ns]

Sample dates:
0   2025-02-25
1   2025-10-14
2   2025-05-13
3   2025-12-02
4   2025-11-20
5   2025-12-31
6   2025-10-14
7   2025-12-05
8   2025-11-21
9   2025-04-16
Name: Order_Date, dtype: datetime64[ns]

Number of unique dates: 342


### Date Data Quality Assessment

The `Order_Date` column is already stored in the `datetime64[ns]` format. The sample values follow a consistent date representation, and the dataset contains 342 unique order dates. Therefore, no additional date-format conversion is required during the cleaning stage.

In [55]:
print("Earliest Order Date:", df["Order_Date"].min())
print("Latest Order Date:", df["Order_Date"].max())
print("Missing Order Dates:", df["Order_Date"].isnull().sum())

Earliest Order Date: 2025-01-01 00:00:00
Latest Order Date: 2026-01-01 00:00:00
Missing Order Dates: 0


###  Check numerical columns for potential outliers

In [56]:
# Check numerical columns for potential outliers

numerical_columns = [
    "Age",
    "Quantity",
    "Unit_Price",
    "Total_Sales"
]

for column in numerical_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ][column].count()
    
    print(f"{column}:")
    print(f"  Lower bound: {lower_bound:.2f}")
    print(f"  Upper bound: {upper_bound:.2f}")
    print(f"  Potential outliers: {outliers}")
    print()

Age:
  Lower bound: -6.00
  Upper bound: 90.00
  Potential outliers: 0

Quantity:
  Lower bound: -4.50
  Upper bound: 15.50
  Potential outliers: 0

Unit_Price:
  Lower bound: -21529.27
  Upper bound: 72937.37
  Potential outliers: 0

Total_Sales:
  Lower bound: -187917.74
  Upper bound: 438707.26
  Potential outliers: 19



### Outlier Assessment

The IQR method identified 19 potential outliers in the `Total_Sales` column. No potential outliers were identified in Age, Quantity, or Unit_Price. The identified Total_Sales values were retained because they are consistent with the relationship `Quantity × Unit_Price` and there was no evidence that they represented data-entry errors.

### View records with missing Age

In [69]:
missing_age = df[df["Age"].isnull()]
missing_age[["Customer_ID","Customer_Name","Age","Gender","City","Product","Category"]]

,Customer_ID,Customer_Name,Age,Gender,City,Product,Category


In [58]:
age_median = df["Age"].median()
print("Median Age:", age_median)

Median Age: 41.0


In [59]:
df["Age"] = df["Age"].fillna(age_median)

In [60]:
print("Missing Age after cleaning:", df["Age"].isnull().sum())
print("Age data type:", df["Age"].dtype)

Missing Age after cleaning: 0
Age data type: float64


### Handling Missing Age Values

There were 20 missing values in the `Age` column. Since the actual ages could not be determined from the available dataset, the missing values were replaced using the median age of 41. Median imputation was selected because it is less sensitive to extreme values and preserves all transaction records.

### Replace missing City values with "Unknown"

In [62]:
df["City"] = df["City"].fillna("Unknown")

In [63]:
print("Missing City after cleaning:", df["City"].isnull().sum())
print("\nCity value counts:")
print(df["City"].value_counts())

Missing City after cleaning: 0

City value counts:
City
Patna        135
Kolkata      133
Mumbai       131
Delhi        125
Hyderabad    125
Bengaluru    122
Gaya         117
Pune          99
Unknown       13
Name: count, dtype: int64


### Handling Missing City Values

There were 13 missing values in the `City` column. Since the actual cities could not be determined from the available data, the missing values were replaced with `Unknown` rather than assigning an assumed city. This preserves the records while clearly identifying the unavailable location information.

### Check missing values after cleaning


In [66]:
missing_values = df.isnull().sum()
print(missing_values)
print("\nTotal missing values:", df.isnull().sum().sum())

Order_ID         0
Order_Date       0
Customer_ID      0
Customer_Name    0
Age              0
Gender           0
City             0
Product          0
Category         0
Quantity         0
Unit_Price       0
Total_Sales      0
dtype: int64

Total missing values: 0


### Check repeated Order IDs

In [67]:
duplicate_order_ids = df[df["Order_ID"].duplicated(keep=False)].sort_values("Order_ID")
print("Number of records with repeated Order_IDs:",duplicate_order_ids.shape[0])
print("\nRepeated Order IDs:")
print(duplicate_order_ids["Order_ID"].value_counts())

Number of records with repeated Order_IDs: 9

Repeated Order IDs:
Order_ID
ORD100050    9
Name: count, dtype: int64


### Duplicate Order_ID Assessment

The dataset contains one repeated Order_ID, `ORD100050`, which appears in 9 records. Although this creates 8 duplicate occurrences, the records contain different transaction details. Therefore, they were retained rather than removed, as the repeated Order_ID does not represent exact duplicate rows.

In [70]:
df[df["Order_ID"] == "ORD100050"]

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,41.0,Male,Pune,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,34.0,Male,Kolkata,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,23.0,Male,Bengaluru,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,24.0,Female,Pune,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,64.0,Male,Bengaluru,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,41.0,Male,Unknown,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,31.0,Female,Mumbai,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,35.0,Male,Kolkata,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,62.0,Male,Bengaluru,Mobile,Electronics,1,34619.77,34619.77


### Duplicate Order_ID Final Decision

The repeated `Order_ID` `ORD100050` appears in 9 records. Detailed inspection showed that these records contain different dates, customers, products, quantities, prices, and sales values. Therefore, they represent different transaction records rather than exact duplicates. The records were retained, and the repeated Order_ID issue was documented instead of removing valid transactions.

In [71]:
text_columns = ["Order_ID","Customer_ID","Customer_Name","Gender","City","Product","Category"]
for column in text_columns:
    df[column] = df[column].astype(str).str.strip()

In [72]:
print(df[text_columns].isnull().sum())

Order_ID         0
Customer_ID      0
Customer_Name    0
Gender           0
City             0
Product          0
Category         0
dtype: int64


In [73]:
print(df.dtypes)

Order_ID                 object
Order_Date       datetime64[ns]
Customer_ID              object
Customer_Name            object
Age                     float64
Gender                   object
City                     object
Product                  object
Category                 object
Quantity                  int64
Unit_Price              float64
Total_Sales             float64
dtype: object


### Convert Age to Integer

In [74]:
df["Age"] = df["Age"].astype(int)
print(df["Age"].dtype)
print(df["Age"].head())

int64
0    30
1    63
2    62
3    65
4    44
Name: Age, dtype: int64


### Post-Cleaning Validation

In [78]:
print("Dataset shape:", df.shape)

print("Total missing values:", df.isnull().sum().sum())

print("Exact duplicate rows:", df.duplicated().sum())

print(df.dtypes)

df.head(10)

Dataset shape: (1000, 12)
Total missing values: 0
Exact duplicate rows: 0
Order_ID                 object
Order_Date       datetime64[ns]
Customer_ID              object
Customer_Name            object
Age                       int64
Gender                   object
City                     object
Product                  object
Category                 object
Quantity                  int64
Unit_Price              float64
Total_Sales             float64
dtype: object


,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,ORD100002,2025-02-25,CUST5529,Customer_227,30,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,ORD100003,2025-10-14,CUST3127,Customer_182,63,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,ORD100004,2025-05-13,CUST8887,Customer_487,62,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,ORD100005,2025-12-02,CUST2515,Customer_470,65,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,ORD100006,2025-11-20,CUST4796,Customer_380,44,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90
5,ORD100007,2025-12-31,CUST3952,Customer_71,40,Female,Hyderabad,Laptop,Electronics,9,35947.47,323527.23
6,ORD100008,2025-10-14,CUST1758,Customer_335,61,Male,Patna,Mobile,Electronics,6,49997.53,299985.18
7,ORD100009,2025-12-05,CUST9622,Customer_224,35,Male,Kolkata,Laptop,Electronics,3,9488.83,28466.49
8,ORD100010,2025-11-21,CUST9770,Customer_494,41,Female,Mumbai,Book,Education,6,19488.23,116929.38
9,ORD100011,2025-04-16,CUST1312,Customer_40,30,Male,Kolkata,Mobile,Electronics,8,18024.34,144194.72


## Final Post-Cleaning Validation

The cleaned dataset was validated after completing the data cleaning and transformation steps. The dataset contains 1,000 rows and 12 columns with no remaining missing values or exact duplicate rows.

The `Order_Date` column is stored in `datetime64[ns]` format, while `Age` and `Quantity` are stored as integers. The `Unit_Price` and `Total_Sales` columns are stored as floating-point values. These checks confirm that the dataset is properly structured and ready for further analysis.

###  Export cleaned dataset

In [82]:
df.to_csv("cleaned_sales_data_ApexPlanet.csv", index=False)

print("Cleaned dataset exported successfully!")
print("File name: cleaned_sales_data_ApexPlanet.csv")

Cleaned dataset exported successfully!
File name: cleaned_sales_data_ApexPlanet.csv


### Check Whether EXISTS : TRUE

In [81]:
import os
print(os.path.exists("cleaned_sales_data.csv"))

True


# Data Cleaning Summary

The dataset was assessed and cleaned to improve data quality and prepare it for analysis.

### Key Data Quality Issues Identified
- 20 missing values were found in the `Age` column.
- 13 missing values were found in the `City` column.
- No exact duplicate rows were found.
- A repeated `Order_ID` (`ORD100050`) was identified, but the associated records contained different transaction details, so they were retained.
- No extra spaces or obvious categorical formatting inconsistencies were identified.
- The `Order_Date` column was already in the appropriate datetime format.
- 19 potential statistical outliers were identified in `Total_Sales` using the IQR method. These were retained because they were mathematically consistent with `Quantity × Unit_Price`.
- The `Total_Sales` values were validated against `Quantity × Unit_Price`, and all 1,000 records matched.

### Cleaning Actions Performed
- Missing `Age` values were replaced with the median age of 41.
- Missing `City` values were replaced with `Unknown`.
- Leading and trailing spaces were removed from text columns.
- `Age` was converted to integer format.
- Date and numerical data types were validated.
- Exact duplicate rows were checked and none were removed.
- Repeated `Order_ID` records were retained after detailed inspection.

### Final Dataset Status

After cleaning and validation, the dataset contains 1,000 rows and 12 columns with zero missing values and zero exact duplicate rows. The dataset is now ready for analysis.

# Data Dictionary

| Column Name | Description | Data Type | Business Relevance |
|---|---|---|---|
| Order_ID | Transaction/order reference identifier | Object | Used to identify and track transaction records |
| Order_Date | Date on which the transaction occurred | Datetime | Useful for analyzing sales over time |
| Customer_ID | Identifier assigned to the customer | Object | Used for customer-level analysis |
| Customer_Name | Name or label associated with the customer | Object | Provides customer reference information |
| Age | Age of the customer in years | Integer | Useful for demographic and customer segmentation analysis |
| Gender | Gender category of the customer | Categorical | Useful for demographic analysis |
| City | Customer's city or location | Categorical | Useful for geographic analysis |
| Product | Product purchased by the customer | Categorical | Used for product-level sales analysis |
| Category | Category to which the product belongs | Categorical | Useful for category-level sales analysis |
| Quantity | Number of units purchased | Integer | Used to measure sales volume |
| Unit_Price | Price of one unit of the product | Float | Used for pricing and sales calculations |
| Total_Sales | Total value of the transaction | Float | Used to analyze sales/revenue performance |